hopper/seq_len.h

```diff
@@ -4,15 +4,25 @@
 
 #pragma once
 
+#include <array>
+#include <algorithm>
+
 #include <cutlass/cutlass.h>
 #include <cute/layout.hpp>
 
 namespace flash {
 
 static constexpr int kMaxTileSize = 128;
 
-template <bool UseVarSeqLen> class SeqLenTraits {
+static constexpr int FixedSeqLenType = 0;
+static constexpr int VarSeqLenType = 1;
+static constexpr int DecodingGQASeqLenType = 2;
+
+template <int SeqLenType, bool Is_dynamic_> class SeqLenTraits {
```

Comment:
What does `Is_dynamic_` mean here? How is it different from var-seq-len?

Reply:
I meant to include a template parameter to guard against reading `actual_seq_len` when not needed as that degrades fp8 kernel perf. Actual seqlen can differ per batch with both fixed or variable seq len. Do you have a better name for this?

Reply:
Why could actual seqlen differ per batch for fixed seq len?

Reply:
I'm using `FixedSeqLenTraitsDynamic` in the kv cache setup where actual seqlen is <= cache seqlen. I only need actual seqlen to be updated every batch and don't need the other layout changes. However, I suppose `VarSeqLenTraits` can be used for this instead -- let me look into it.

Reply:
In the kv cache setup I don't have access to `total_k`, so won't be able to use `VarSeqLenTraits` layouts.

Reply:
In the latest update I removed `Is_dynamic_` entirely. I found another way to handle perf regression with the fp8 kernel, through introducing another init method `init_no_guard` that involves no conditioning on the pointer not being nullptr internally, but instead moving that logic outside and using a boolean in smem instead.